app/models/nic/base.rb

```diff
@@ -202,6 +204,19 @@ def sync_name
       end
     end
 
+    def validate_host_taxonomy
+      return true if self.subnet.nil?
+      return true if self.host.organization.nil? && self.subnet.organizations.empty? &&
+                     self.host.location.nil? && self.subnet.locations.empty?
+
+      errors.add(:subnet, _("is not defined for host's organization.")) unless include_or_empty?(self.subnet.organizations, self.host.organization)
+      errors.add(:subnet, _("is not defined for host's location.")) unless include_or_empty?(self.subnet.locations, self.host.location)
+    end
+
+    def include_or_empty?(list, item)
+      return true if list.empty? and item.nil?
```

Comment:
It's better to use `&&` in conditions (different priority than `and`), `and` is rather used for program flow control